SassTransformer should be an AggregateTransformer

luisvt opened the pull request:
Making SassTransformer extends AggregateTransformer allow us to generate one css file from one entry_point listening changes in multiple scss files

For example as user having `main.scss`:

``` scss
@import "variables"

.navbar-default {
  background-color: $default-bg;
}
```

and `variables.scss`:

``` scss
$default-bg: blue;
```

We want to be able to:
- parse and generate only main.css
- edit variable.scss and the transformer should regenerate main.scss

komu commented:
Hi! Thanks for your PR!

The original code already works in the same way as Sass does: if your file name starts with an underscore, it's not translated to CSS, but can still be included. Check out [section on Partials](http://sass-lang.com/guide#topic-4) in Sass Guide. The nice thing about this mechanism is that it's already understood by all Sass-aware tools and requires to extra configuration.

So if your `main.scss` is exactly as in your example, but you simply rename `variables.scss` to `_variables.scss`, everything should work as you want it to work: changes to either `main.scss` or `_variables.scss` will rebuild the `main.css`, but no `variables.css` is ever generated.

So it might be possible to simplify the old code using the new `AggregateTransformer`-class, but the functionality is already there.

Anyway, thanks for your PR, but I'm not merging this. (Unless I'm completely missing something here?)

luisvargastije committed:
changes:

- Convert `SassTransformer` into aggregate transformer, this reduce the complexity of the code and avoid wrongly parsing `@import` inside comments (fixes #6)
- Removed unneeded files
  *  `base_sass_transformer.dart`
  * `sass_file.dart`
  * `inlined_sass_file.dart`
  * `inlined_sass_transformer.dart`
  * `run.sh`
  * `inlined_sass_transformer_test.dart`
  * `sass_file_test.dart`

zoechi commented on README.md:
````diff
@@ -16,59 +18,67 @@ transformers:
 ```
 
-After adding the transformer your `.sass` and `.scss` files will be automatically transformed to
+After adding the transformer all your `.sass` and `.scss` files will be automatically transformed to
````
I think there should be a `,` after "After adding the transformer". I started reading this sentence a few times until I got what it means ;-)

zoechi commented on README.md:
````diff
@@ -16,59 +18,67 @@ transformers:
 ```
 
-SassC only supports `.scss`-files and does not support Compass.
+> Doing `pub serve` during development, css files are going to live in memory only. When you do `pub build` actual css files will be output in build folder
````
I would change this to "By using `pub serve` ...". "Executing `pub build` creates actual css files in the build folder.

zoechi commented on lib/src/transformer_options.dart:
```diff
@@ -1,26 +1,69 @@
+/// Class to be used to parse sass transformer options coming from pubspec.yaml file
 class TransformerOptions {
+  /// entry_point: web/builder.sass - main file to build or [file1.sass, ...,fileN.sass]
+  final List<String> entry_points;
```
field names shouldn't contain underscores they should be lowerCamelCase

zoechi commented:
@luisvt AFAIK what @komu wrote in https://github.com/dart-league/dart-sass/pull/2#issuecomment-171892478 about partials is legit. 
I'm not sure I understand what problem this PR actually solves.

zoechi commented:
I'm also wondering if the changes might break existing users.

luisvargastije committed:
changes:

- Rename `entry_points` to `entryPoints` and `include_paths` to includePaths into `transformer_options.dart`
- Add corrections suggested by @zoechi in readme

luisvt commented:
@zoechi you are right, what @komu wrote is correct.

> The original code already works

That's right the original code already works, however it produces some bugs like:
- [`@imports` inside comments being processed](https://github.com/dart-league/dart-sass/issues/6)
- [weird pub get fails](https://github.com/dart-league/ng_bootstrap/issues/20)
- entry point files will need to use full package url like this: https://github.com/dart-league/bootstrap_sass/blob/1043992c9ad7ef006847f3a5b6f4ef332652f921/lib/scss/bootstrap.scss#L9

For the second question, right now changes will brake some existing users. Instead users writing:

``` yaml
dependencies:
  sass: any
transformers:
  - sass
```

they will need to write:

``` yaml
dependencies:
  sass: any
transformers:
  - sass:
      entry_points: some_path/main.css
      include_paths: some_path
```

zoechi commented:
> @imports inside comments being processed

This means that imports need to be maintained twice, in the import path and in the transformer configuration (only once for every directory), right?

> entry point files will need to use full package url like this: https://github.com/dart-league/bootstrap_sass/blob/1043992c9ad7ef006847f3a5b6f4ef332652f921/lib/scss/bootstrap.scss#L9

How does this work with the PR? Does this allow you to use relative paths to scss files within the current package?
How do you import SCSS files from dependency packages?
Does this work with `pub get --no-packages-dir`?

I don't want to be pedantic. It's great you put so much effort into this work!
I just try to understand the implications.

luisvargastije committed:
changes:

- avoid processing sass files starting with `_`
- include path of processable assets. This allow us to skip adding the asset path manually

luisvt commented:
Let me check. Since I'm using an aggregate transformer, it still should listen on changes of all `.scss` and `.sass` files, but only process main files.

luisvt commented:
The transformer is working as expected, only main files (ie. files that don't start with `_` will be processed by transformer.

Aggregate transformer listen on changes of all `primary assets` (in this case all files with extension `sass` or `scss`) see: https://github.com/luisvt/dart-sass/blob/master/lib/transformer.dart#L28

Whenever a change occurs in any `scss` or `sass` file, then the apply function will run for all the `primary assets` (ie. all `scss` and/or `sass` files). What I do in these lines: https://github.com/luisvt/dart-sass/blob/master/lib/transformer.dart#L38-L42 is I skip processing files that begins with `_`. So that, files that doesn't start with `_` will be processed.

In few words, the aggregate transformer code will work like @komu's code, without all the complexity.

luisvargastije committed:
remove `entry_points` option

luisvargastije committed:
Update README and Docs:

* Remove `etnry_points` from README
* Add configurations to README
* Enhance documentation to `TransformerOptions`

zoechi commented on README.md:
````diff
@@ -18,38 +18,45 @@ transformers:
 ```
 
-After adding the transformer, all your `.sass` and `.scss` files will be automatically transformed to
+After adding the transformer, all your `.sass` and `.scss` files that doesn't begin with `_` will be automatically transformed to
````
I guesss

```
that doesn't begin with
```

should be 

```
that don't begin with
```

zoechi commented on README.md:
```diff
@@ -18,38 +18,45 @@ transformers:
 
-If your main file imports other files you will need to add the option `include_paths`,
- to let `sass` know which folder will be used for processing imports:
+If your main file imports other files outside the main files folder, you need to add the option `include_paths`,
```
does this mean `lib` needs to be added when the main file is in `web`?

luisvt replied:
Nop, that means that if `web/main.scss` has:

``` scss
@import "partial";
```

and partial is in `lib/_partial.scss` it will be needed to add `include_paths` option.

``` yaml
- sass:
    include_path: lib
```

If `web/main.scss` has:

``` scss
@import "lib/partial"
```

or:

``` scss
@import "package/<package-name>/partial"
```

it wouldn't be needed to add `include_paths` parameter.

komu commented:
Well, first of all there is a simple bug in your `classifyPrimary` that you'll probably want to fix. The code is:

``` dart
classifyPrimary(AssetId id) =>
    ['.scss', '.sass'].any((e) => e == id.extension) ? id.extension : null;
```

This is wrong because it classifies `.scss` and `.sass` files into different buckets. Therefore if  a `.sass` file imports `.scss` file, the changes to the `.scss` file will not be picked up. If you'd like to pursue this route, you'd want to put all assets into one bucket instead of separating them by extension.

However, that brings us to our second problem: if you just blindly put all assets into one bucket regardless of their actual dependency chain, it will rebuild everything when anything is changed. While with `sassc` the actual Sass transformation time is negligible, if the outputs are fed to other transformers, it will cause extra work. A common simple scenario works by brute force, but causes problems for more complex setups.

Finally, I'd like bring you attention to [inlined_sass_transformer](https://github.com/dart-league/dart-sass#inlined-transformer) and the [issue related to that](https://bitbucket.org/evidentsolutions/dart-sass/issues/4/support-transformations-to-imported-sass). The complexity of going through the files and processing imports is not really going to go away unless you want to break that (perfectly valid) workflow. In fact, in the long run it would make sense to make all transformation use that approach, but it would need work on remapping the error messages.

So as a summary: the "just bruce force and let sass do everything" approach works for simple setups where you don't have transformers before or after sass, but unfortunately not all setups are that simple. I, more than anyone, would be happy to purge obsolete code and make programs simpler where possible, but here it seems dubious if that is really possible. Such is life. :)

zoechi commented:
@komu

> in the long run it would make sense to make all transformation use that approach,

Do you mean this approach? "going through the files and processing imports"

It also seems to be necessary to rewrite imports if they start with `packages` when the `packages` folder goes away which could happen pretty soon.
Another approach would be to keep generating symlinks for SASS imports to keep them working :-/

zoechi commented:
> It also seems to be necessary to rewrite imports if they start with packages when the packages folder goes away which could happen pretty soon.

I don't have any idea yet how that could work, because SASS would need to see these changes.

komu commented:
Yes, exactly. In the long run a nice solution would be to inline all imports and pass a pre-processed file to sass. That would solve problems with removing `packages`, with transformer chains, etc. So essentially an opposite approach to this PR.

It would need some on the parsing side (at least adding support for comments) and it would a really nice thing to map the errors back to original locations. So it would definitely mean more work `dart-sass`, but then everything would "just work" for the user.

luisvt commented:
> Well, first of all there is a simple bug in your classifyPrimary that you'll probably want to fix. The code is:
> 
> ``` dart
> classifyPrimary(AssetId id) =>
>     ['.scss', '.sass'].any((e) => e == id.extension) ? id.extension : null;
> ```
> 
> This is wrong because it classifies .scss and .sass files into different buckets.

There is no such bug, they always get processed in the same bucket.

> However, that brings us to our second problem: if you just blindly put all assets into one bucket regardless of their actual dependency chain, it will rebuild everything when anything is changed. While with sassc

Only main `sass` or `scss` files will get processed. Even though everything is built again, the time spent is around 2 seconds at the most.

>  [inlined_sass_transformer](https://github.com/dart-league/dart-sass#inlined-transformer) and the [issue related to that](https://bitbucket.org/evidentsolutions/dart-sass/issues/4/support-transformations-to-imported-sass). The complexity of going through the files and processing imports is not really going to go away unless you want to break that (perfectly valid) workflow.

Yes, for that scenario, there is no way to get rid of processing imports. So that, I reverted back the deletion of `inline_sass_transformer`. So that, we could still have both transformers working.

luisvargastije committed:
changes:

* Correct typo in Readme
* add back inlined_sass_transformer
* make sass_transformer don't require users to add `include_path`

komu commented:
Luis, it feels like you are operating under the assumption that I'm an idiot who throws around completely unfounded ideas that can be ignored without further consideration. This makes communication pretty tiring. I understand that it's very hard to communicate with strangers in the internet, but let's try anyway.

I've gone out of my way to read and understand your code, install Dart again and actually test your code in different scenarios. I'd hope that you'd spend some effort in trying to understand what I'm trying to say. I've probably not worded all the things as well as I could, but if it feels like I'm saying something that clearly makes no sense, please consider the option that you don't understand the big picture.

With this in mind, I'd like you to read the previous message again.

Done? I'll continue now:

> There is no such bug, they always get processed in the same bucket.

There indeed is such a bug: it's clearly visible in the code because it returns `id.extension` or `null` from `classifyPrimary`. This means that there are going to be two buckets: one named `.scss` and other named `.sass`. Whenever a file in the bucket changes,  all files in the same bucket are means. This means that if you have `foo.scss` importing `_bar.sass` and you make a change to `_bar.sass`, it won't be picked up automatically.

The bug is perfectly visible the code I quoted, but because I don't want to waste anyone's time reporting bugs that I have not verified, I also tested the behavior before writing the previous message.

There is a bug in the code.

> Only main sass or scss files will get processed.

Perhaps I was a little unclear. Let my try again:

I fully understand what your code does and even said that the time of Sass-transformation is negligible, but _"if the outputs are fed to other transformers, it will cause extra work."_ Your own builds probably don't have several built CSS-files that are _further processed by other more expensive transformers_ after Sass has built them, but it's a perfectly valid build setup. Not having the real dependency chain can cause non-trivial amount of extra work performed.

This is not a huge problem in itself: it could be worthwhile to have performance regressions for more exotic cases if it makes the code simpler and more maintainable. However, coupled with the next issue and the ideas that I had about the future directions that would be worthwhile to pursue, it doesn't seem like this would be the case.

> Yes, for that scenario, there is no way to get rid of processing imports. So that, I reverted back the deletion of inline_sass_transformer. So that, we could still have both transformers working.

Right, as you probably noticed, most of the code between normal transformer and inlined transformer is in fact shared in the `BaseSassTransformer`. The default transformer adds very little complexity on top of that.

So if we'd merge this PR, we'd have two completely different ways to trigger the transformation and this would not in fact reduce complexity, but add it. Furthermore, there would be two separate code paths having completely different failure modes.

Finally there's the question of future: well behaving transformers should work even if there are no files written to disk at all. (This is related both to `pub`-maintainers' hopes to get rid of `packages` folders and to transformer chains where the inputs of previous transformers are only cached in memory.)

One of the main reasons why I didn't merge this PR previously was not that it wouldn't be a bad solution for status quo, but that it moves the code towards wrong direction in terms of what is needed for future. We really need to use Barback's APIs more and not less to get all the required files. (Of course all that context was hidden inside my head and in comments on various issues, so you had no way of knowing all the forces relating to this issue.)

So: if inlined transformer would map error messages correctly, it would be superior in every respect and I'm pretty sure that improving that is the right way to proceed. Since the default transformer uses the same code, the improvements will automatically benefit that as well. For now, there are downsides in using the inlined transformer (error messages are not mapped to their original locations), but once those are fixed, the distinction between transformers could be removed.

luisvt commented:
First of all, sorry. In any moment I did not make the assumption that you are stupid. And Secondly, This conversation is not going anywhere, so I will close this PR and create a new separate transformer. When the issues/bugs mentioned before get corrected I will comeback to this transformer. Thanks.
